WIP: sixad: improve coexistence of first & third-party controllers

psyke83 opened the pull request:
Remove GASIA_GAMEPAD_HACKS macro and use the appropriate
configuration based on each detected controller instead.

--

Don't merge; needs testing by owners of third-party controllers such as @pjft.

The intent of this PR is to eliminate the need for separate branches, so when testing this PR, please install using the first option of the scriptmodule (so it fetches the master branch, that this PR affects). If the patch works well, I'll send a PR to clean up the scriptmodule afterwards.

joolswills commented:
Thanks for this - it will be much more manageable with a single branch to maintain.

psyke83 commented on bluetooth.cpp:
```diff
@@ -203,16 +204,15 @@ void l2cap_accept(int ctl, int csk, int isk, int debug, int legacy)
-    req.product = 0x0268;
-    req.version = 0x0100;
-    req.parser  = 0x0100;
+    if (get_sdp_device_info(&addr_src, &addr_dst, &req) < 0) {
```
I'm uncertain if this function will actually fail on a Gasia controller. I know that third-party Shanwan controllers fail to pair on BlueZ due to a missing SDP record (which is patched in newer releases: https://git.kernel.org/pub/scm/bluetooth/bluez.git/commit/plugins/sixaxis.c?id=1629c39ededef07988a5403b27331e0e317f1e08).

Additionally, this may cause Shanwan controllers to use the Gasia configuration, which may be unwanted. If at least req.name is enumerated by get_sdp_device_info by a Shanwan controller, and doesn't fake the official "Sony Computer Entertainment Wireless Controller" string, then we could do a string compare against "SHANWAN" etc. to see if we really want to set gasia = 1 here.

psyke83 commented:
I updated the PR:
* added a small function that queries the remote name of the controller to help detect the difference between official, gasia and shanwan controllers.
* get_sdp_device_info() is called only for official controllers, and the manual configuration fallback is used for both shanwan and gasia.
* the gasia-specific configuration will no longer be used on Shanwan controllers (not sure if this is the ideal behaviour until it's tested)

To help troubleshoot, the function will print the name of the controller in the log, like so:
```
Jan 11 01:57:09 retropie-test sixad-bin[19566]: remote name: PLAYSTATION(R)3 Controller
```

Will add some additional notes inline.

psyke83 commented on sixaxis.cpp:
```diff
@@ -432,27 +428,20 @@ void do_rumble(int csk, int led_n, int weak, int strong, int timeout)
-#endif
+    if (!gasia) {
+        unsigned char buf[128];
+        recv(csk, buf, sizeof(buf), 0); //MSG_DONTWAIT?
```
The recv command may break shanwan. It may also be safe to simply remove for all controllers, but I need to test that. If it breaks shanwan, then detection for shanwan can also be added here.

psyke83 commented on sixaxis.cpp:
```diff
@@ -549,15 +537,13 @@ int set_sixaxis_led(int csk, struct dev_led led, int rumble)
-    recv(csk, buf, sizeof(buf), 0);
-#endif
+    if (!gasia)
+        recv(csk, buf, sizeof(buf), 0);
```
Same as previous comment (may break shanwan & may also not be needed for official controller).

psyke83 commented:
OK, updated the PR to allow shanwan to be exported via arguments and restricted the recv commands to official controllers only. Apologies for the spam; will leave the PR alone until I get feedback.

pjft commented:
Not at all, thanks for all the work here and for the updates. I didn't have time to check this yesterday but will try to do it this evening. 

Have a great day!

psyke83 commented:
Closing this as the backported hid-sony & BlueZ is a preferable solution.
